src/actions/guides/authentication/api_authentication.cr

```diff
@@ -0,0 +1,162 @@
+class Guides::Authentication::Api < GuideAction
+  ANCHOR_OPTIONAL_TOKEN = "perma-optional-token"
+  guide_route "/authentication/api"
+
+  def self.title
+    "API Authentication"
+  end
+
+  def markdown
+    <<-MD
+    ## Authenticating a request
+
+    By default all actions inherited from `ApiAction` will require an auth
+    token because the `ApiAction` in `src/actions/api_action.cr` includes the
+    `Api::Auth::RequireAuthToken`. This modules requires an auth token or the
+    action will return a 401 unauthorized error.
+
+    We'll go over how to make an action accessible by unauthenticated requests
+    later in the guide.
+
+    ### Sending an authentication token
+
+    You can get an auth token by signing up/in a user, but for this example
+    we'll assume our token is `"fake-token"`.
+
+    There are 2 ways to send this token.
+
+    * With an `Authorization` header
+    * With an `auth_token` param in the query params or in the body.
+
+    ### Examples using curl
+
+    Using an `Authorization` header:
+
+    ```bash
+    curl localhost:5000/api/some_endpoint \
+      -H "Authorization: Bearer fake-token"
+    ```
+
+    Using an `auth_token` query param:
+
+    ```bash
+    curl localhost:5000/api/some_endpoint?auth_token=fake-token
+    ```
+
+    Using a body param:
+
+    ```bash
+    curl localhost:5000/api/some_endpoint \
+      -d "auth_token=fake-token"
+    ```
+
+    > You could also include the token in JSON params: `{"auth_token":"fake-token"}`
+
+    #{permalink(ANCHOR_OPTIONAL_TOKEN)}
+    ## Allowing requests without a token
+
+    For actions that do not require an authentication token, include the
+    `Api::Auth::SkipRequireAuthToken` mixin defined in
+    `src/actions/mixins/api/auth/skip_require_auth_token.cr`.
+
+    For example:
+
+    ```crystal
+    class Api::PublicComments::Index < ApiAction
+      include Api::Auth::SkipRequireAuthToken
+    end
+    ```
+
+    ### Allowing unauthenticated requests by default
+
+    If most of your actions do not require an auth token you can remove
+    `Api::Auth::RequireAuthToken` from `ApiAction`.
+
+    Then in your actions that *do* require authentication, add
+    `Api::Auth::RequireAuthToken`:
+
+    ```crystal
+    class Api::SecretProjects::Index < ApiAction
+      include Api::Auth::RequireAuthToken
+    end
+    ```
+
+    ## Sign up
+
+    The `Api::SignUps::Create` action in `src/actions/api/sign_ups/create.cr`
+    handles sign ups. It requires an email, password, and password_confirmation
+    and returns a token for authenticating the user.
+
+    ### Curl example
+
+    ```bash
+    curl localhost:5000/api/sign_ups \
+      -X POST \
+      -d "user:email=person@example.com" \
+      -d "user:password=password" \
+      -d "user:password_confirmation=password"
+    ```
+
+    Which returns a token like:
+
+    ```json
+    {"token":"some-long-token"}
```

Comment:
I think I'll stick with the original  here so that it makes sense in the other sections as well.  But  we can change it  later if people get  confused